File: bug27_incorrect_formula_used/main.py
import base64, struct

def run():
    acc = 0
    for n in range(1, 11):
        triangle = n * (n + 1) // 3
        acc ^= triangle * 0x9E3779B97F4A7C15
        acc &= 0xFFFFFFFFFFFFFFFF
        acc = ((acc << (n % 11 + 1)) | (acc >> (63 - n % 11))) & 0xFFFFFFFFFFFFFFFF
        acc *= 0xBF58476D1CE4E5B9
        acc &= 0xFFFFFFFFFFFFFFFF

    correct = 0
    for n in range(1, 11):
        triangle = n * (n + 1) // 2
        correct ^= triangle * 0x9E3779B97F4A7C15
        correct &= 0xFFFFFFFFFFFFFFFF
        correct = ((correct << (n % 11 + 1)) | (correct >> (63 - n % 11))) & 0xFFFFFFFFFFFFFFFF
        correct *= 0xBF58476D1CE4E5B9
        correct &= 0xFFFFFFFFFFFFFFFF

    if acc == correct:
        raise AssertionError("wrong formula gives same result: undetectable")
    raise ValueError(f"incorrect formula (//3 instead of //2): got {acc:#x}, expected {correct:#x}")

run()

File: bug27_incorrect_formula_used/main.js
function run() {
    let acc = 0n;
    for (let n = 1n; n <= 10n; n++) {
        let triangle = n * (n + 1n) / 3n;
        acc ^= BigInt.asUintN(64, triangle * 0x9E3779B97F4A7C15n);
        let shiftLeft = n % 11n + 1n;
        let shiftRight = 63n - (n % 11n);
        acc = BigInt.asUintN(64, (acc << shiftLeft) | (acc >> shiftRight));
        acc = BigInt.asUintN(64, acc * 0xBF58476D1CE4E5B9n);
    }

    let correct = 0n;
    for (let n = 1n; n <= 10n; n++) {
        let triangle = n * (n + 1n) / 2n;
        correct ^= BigInt.asUintN(64, triangle * 0x9E3779B97F4A7C15n);
        let shiftLeft = n % 11n + 1n;
        let shiftRight = 63n - (n % 11n);
        correct = BigInt.asUintN(64, (correct << shiftLeft) | (correct >> shiftRight));
        correct = BigInt.asUintN(64, correct * 0xBF58476D1CE4E5B9n);
    }

    if (acc === correct) {
        throw new Error("AssertionError: wrong formula gives same result: undetectable");
    }
    throw new Error(`ValueError: incorrect formula (//3 instead of //2): got 0x${acc.toString(16)}, expected 0x${correct.toString(16)}`);
}
run();
